Provide default implementations of AbstractStore log methods

acroz opened the pull request:
The abstract store has methods for logging individual metrics, params and tags, however since these are not marked as abstract methods they're not currently required to be implemented in tracking store implementations.

This change provides default implementations of `log_metric`, `log_param` and `set_tag` that proxy calls to the more general `log_batch` added by @smurching in #950.

Following this change, as long as stores implement `log_batch`, the above methods need not be provided.

acroz committed:
Provide default implementations of AbstractStore log methods

The abstract store has methods for logging individual metrics, params
and tags, however since these are not marked as abstract methods they're
not currently required to be implemented in tracking store
implementations.

This change provides default implementations of `log_metric`,
`log_param` and `set_tag` that proxy calls to the more general
`log_batch` added by @smurching in #950.

Following this change, as long as stores implement `log_batch`, the
above methods need not be provided.

mparkhe commented:
This is a great catch. However, does it make sense to mark all API level methods in abstract store as `@abstractmethod` requiring implementing stores to have to define those and explicitly throw `NotImplemented`?

There are a few more methods in `AbstractStore` that need to be looked at--either marked as `@anstractmethod` or removed.

- `get_experiment_by_name`
- `update_run_info`
- `create_run`
- `list_run_infos`

acroz commented:
Hey, thanks for the comments.

`get_experiment_by_name` I think doesn't need to be set as abstract since it has a default implementation that calls `list_experiments`, which does need to be implemented by subclasses.

The other three should be abstract - I'll update those on this PR.

acroz committed:
Mark other required methods as abstract

acroz committed:
Patch all abstract methods

To make construction of the abstract class work

acroz committed:
Forgot to add fixture to test functions

acroz committed:
list_run_infos has a default implementation and so doesn't need to be abstract

acroz committed:
Remove RestStore.list_run_infos

The implementation is the same as that inherited from `AbstractStore`,
so override is redundant

acroz commented:
Correction - `list_run_infos` also has a default implementation, so I've removed its abstract method decorator. I've also removed the redundant identical implementation in `RestStore`.

acroz committed:
Don't need to patch list_run_infos any more

acroz committed:
Ignore unused argument warning with pytest fixture

acroz committed:
Need to subclass to pass Python 2.7 tests

acroz committed:
Fix linting errors

acroz committed:
Remove duplicate method.

acroz commented:
@mparkhe I think I've resolved the outstanding issues on this PR, but the tests are failing for a seemingly unrelated issue with Sagemaker tests. Is this an issue on master?

mparkhe commented:
> @mparkhe I think I've resolved the outstanding issues on this PR, but the tests are failing for a seemingly unrelated issue with Sagemaker tests. Is this an issue on master?

Yes. @smurching and @dbczumar are looking into it.

mparkhe commented on mlflow/store/rest_store.py:
```diff
@@ -224,17 +224,6 @@ def search_runs(self, experiment_ids, search_filter, run_view_type):
         response_proto = self._call_endpoint(SearchRuns, req_body)
         return [Run.from_proto(proto_run) for proto_run in response_proto.runs]
 
-    def list_run_infos(self, experiment_id, run_view_type):
```
:+1:

mparkhe commented on tests/store/test_abstract_store.py:
```diff
@@ -0,0 +1,86 @@
+from mlflow.entities import ViewType
+
+
+class ConcreteStore(AbstractStore):
```
🤣 that name.

Can we rename this to something like `TestAbstractStoreImpl`

acroz replied:
Done - though I do miss the `ConcreteStore` 😁 

I made it `AbstractStoreTestImpl` as classes named `Test*` are inspected by pytest for test methods, I believe.

mparkhe commented on tests/store/test_abstract_store.py:
```diff
@@ -0,0 +1,86 @@
+    def log_batch(self, run_id, metrics, params, tags):
+        raise NotImplementedError()
+
+
```
Would be good to write tests for non-abstract methods in `AbstractStore`. You have covered `log_*` methods--can we write one for `list_run_infos` as well?

acroz replied:
Done - also added tests for `get_experiment_by_name` to complete the set.

mparkhe reviewed (approved):
A couple of small recommendations. LGTM otherwise.

acroz committed:
Rename test store implementation

acroz committed:
Add test for get_experiment_by_name

acroz committed:
Add test for list_run_infos

acroz commented:
@mparkhe Have addressed your comments and now rerunning tests.

acroz committed:
Add missing assertion

acroz commented:
@mparkhe @dbczumar Is there anything left to do to get this merged?

dbczumar reviewed (approved):
LGTM

dbczumar commented:
@acroz Merging now!
